### PR TITLE
Updated load function to optionally load an array of prerendered peak data

### PR DESCRIPTION
**Pull request by rbwest**
Wavesurfer allows for an optional array of peak data to be passed to the load function. This allows the waveform to be generated much quicker for large audio files. I've updated the plugin to accommodate this feature.

### CODE REVIEW

File: src/js/videojs.wavesurfer.js [hunk 1]
```diff
@@ -253,13 +253,31 @@ class Wavesurfer extends Plugin {
      * @param {string|blob|file} url - Either the URL of the audio file,
      *     a Blob or a File object.
      */
-    load(url) {
-        if (url instanceof Blob || url instanceof File) {
-            this.log('Loading object: ' + JSON.stringify(url));
-            this.surfer.loadBlob(url);
+    load(audioUrl, peakUrl) {
+        let player = this;
```
Comment (thijstriemstra) [inline src/js/videojs.wavesurfer.js]: @rbwest can you simply refer to `this` instead of defining and using the `player` var here?

Comment (thijstriemstra) [reply src/js/videojs.wavesurfer.js]: and I also prefer `url` over `audioUrl`. Can you also document the `peakUrl` argument?

File: src/js/videojs.wavesurfer.js [hunk 2]
```diff
@@ -253,13 +253,31 @@ class Wavesurfer extends Plugin {
...
+        if (audioUrl instanceof Blob || audioUrl instanceof File) {
+            player.log('Loading object: ' + JSON.stringify(audioUrl));
+            player.surfer.loadBlob(audioUrl);
         } else {
-            this.log('Loading URL: ' + url);
-            this.surfer.load(url);
+            player.log('Loading URL: ' + audioUrl + '\nLoading Peak Data URL: ' + peakUrl);
+            // Load peak data from file
+            if (peakUrl !== undefined) {
```
Comment (thijstriemstra) [inline src/js/videojs.wavesurfer.js]: there's no need to use the `XMLHttpRequest` here directly, use wavesurfer's https://github.com/katspaugh/wavesurfer.js/blob/master/src/util/ajax.js `ajax` util instead.

File: src/js/videojs.wavesurfer.js [hunk 3]
```diff
@@ -214,7 +214,7 @@ class Wavesurfer extends Plugin {
                 this.player.loadingSpinner.show();
 
                 // start loading file
-                this.load(options.src);
+                this.load(options.src, options.peaks);
```
Comment (thijstriemstra) [inline src/js/videojs.wavesurfer.js]: if you're going to add a new option, can you also add it to the documentation (and default settings)?

File: src/js/videojs.wavesurfer.js [hunk 4]
```diff
@@ -252,14 +252,34 @@ class Wavesurfer extends Plugin {
...
      */
-    load(url) {
+    load(url, peakUrl) {
         if (url instanceof Blob || url instanceof File) {
             this.log('Loading object: ' + JSON.stringify(url));
             this.surfer.loadBlob(url);
         } else {
-            this.log('Loading URL: ' + url);
-            this.surfer.load(url);
+            this.log('Loading URL: ' + url + '\nLoading Peak Data URL: ' + peakUrl);
```
Comment (thijstriemstra) [inline src/js/videojs.wavesurfer.js]: this log message should move into the if-statement below.

**Comment by rbwest** [comment]
@thijstriemstra While working on this pull request I think I've found an issue in videojs-wavesurfer. The loading spinner doesn't seem to be working as intended. When testing on larger audio files the loading spinner does not show up while the waveform is being rendered. Should I start an issue for this?

**Comment by thijstriemstra** [comment]
@rbwest sounds good but I think this is a known issue in a way because; loading large files will cause the browser to slow down to a point that it's not even able to simultaneously render a spinning load icon. This is because loading audio is happening in the same thread as the browser UI. It can be worked around with audio workers but these are not yet available afaik. If you can work around this I'd love to accept a pull request! Cheers.

**Comment by thijstriemstra** [comment]
Thanks!

**Comment by thijstriemstra** [comment]
@rbwest are you able to generate the peaks for `hal.wav`? Took me hours to compile `audiowaveform` for OSX and eventually succeeded but the peaks file generates a weird image.. What params should I use?

I used the default, e.g.

```sh
audiowaveform -i examples/media/hal.wav -o hal.dat -z 256 -b 8
audiowaveform -i hal.dat -o examples/media/hal.json
```

but it's weird output:

screen shot 2017-12-14 at 12 33 01

screen shot 2017-12-14 at 12 50 12

**Comment by jronallo** [comment]
@thijstriemstra I don't know if this answers your issue (and @rbwest is out until the new year) but I believe the numbers that audiowaveform creates are not what wavesurfer.js expects. So we convert the output from audiowaveform into what works in wavesurfer.js. So something like all the numbers being between -1.0..1.0 while audiowaveform uses numbers outside that range. Sorry, but I'm not sure exactly how he's doing the conversion.

So take a look at the floats in the the working data file here: https://avd.lib.ncsu.edu/av/scrc/mc/mc00066-001-bx0001-002-001/mc00066-001-bx0001-002-001-waveform.json
You can see links to to the audio files we have available here: https://avd.lib.ncsu.edu/av/scrc/mc/mc00066-001-bx0001-002-001/data.json

**Comment by thijstriemstra** [comment]
Thanks for the feedback, ill check it out.

**Comment by jronallo** [comment]
This Ruby code seems to do the necessary scaling. So read in the `data` array from the audiowaveform JSON and run it through something like this and you ought to get the scaled data wavesurfer.js requires.

```ruby
def scale_array(arr, domain = [-1, 1])
  min = arr.min
  ratio = (domain[1] - domain[0]).to_f / (arr.max-min).to_f
  arr.map do |x|
    domain[0] + ratio * (x - min)
  end
end
```

**Comment by rbwest** [comment]
@thijstriemstra Sorry for the delay getting back to you. @jronallo is right; wavesurfer.js expects the peak data to be in the range [-1,1]. audiowaveform intentionally leaves scaling to be done at process time. I've included a python file which does this scaling below, as well as a scaled version of hal.wav's peak data. Note that this python file (and the Ruby snippet @jronallo included above) turns all the data values (initially 16 bit) into 32 bit floats, which artificially bloats the file.
[Archive.zip](https://github.com/collab-project/videojs-wavesurfer/files/1564451/Archive.zip)

**Comment by thijstriemstra** [comment]
Thanks that worked. Steps:

1. compile and install https://github.com/bbc/audiowaveform
1. create peaks data file:
```
audiowaveform -i test/support/demo.wav -o test/support/demo.dat -z 256 -b 8
```
3. turn data file into json:
```
audiowaveform -i test/support/demo.dat -o test/support/demo-peaks.json
```
4. fix range using `scale-json.py` in attached https://github.com/collab-project/videojs-wavesurfer/files/1564451/Archive.zip:
```
 python ~/Desktop/scale-json.py test/support/demo-peaks.json
```

`test/support/demo-peaks.json` now contains the correct peaks data.

**Pull request merged**